### PR TITLE
added tests and did some cleanup

### PR DESCRIPTION
**Pull request by sourencho**
+r

### CODE REVIEW

File: src/crawler.js [hunk 1]
```diff
@@ -61,7 +65,7 @@ function normalizeIpp(ip, port) {
 
     out_ip = splitIp
     out_port = port || splitPort || DEFAULT_PORT
-    ipp = out_ip + ':' + out_port
+    var ipp = out_ip + ':' + out_port
   }
 
   return ipp
```
Comment (tdfischer) [inline src/crawler.js:73]: What if `!ip`?

File: src/crawler.js [hunk 2]
```diff
@@ -71,22 +75,43 @@ function normalizeIpp(ip, port) {
...
   EventEmitter.call(this);
-  this.maxRequests = maxRequests ? maxRequests : 30;
+  
+ 
+  // maxRequests checks
+  maxRequests = maxRequests ? maxRequests : 30;
+  if (!maxRequests || maxRequests < 1) {
+    throw new Error("Invalid max requests");
+  }
+  check.assert.number(maxRequests, "Invalid max requests");
```
Comment (tdfischer) [inline src/crawler.js]: This should be moved to the line before `maxRequests < 1`, but after `maxRequests = ...`.

First check makes sure that you've got some default value

Then the code should check if the value passed in is in fact a number.

Finally, check if the number is within the valid range.

File: src/crawler.js [hunk 3]
```diff
@@ -71,22 +75,43 @@ function normalizeIpp(ip, port) {
...
+  
+  // logger checks
+  logger = logger ? logger : console
+  if (!logger.log || !logger.error) {
```
Comment (tdfischer) [inline src/crawler.js:90]: Should be checking `if (typeof logger.log == "function")`. Its entirely possible to pass in an object of `{log: true, error: true}`.

File: src/crawler.js [hunk 4]
```diff
@@ -71,22 +75,43 @@ function normalizeIpp(ip, port) {
...
+    throw new Error("Invalid logger");
```
Comment (tdfischer) [inline src/crawler.js:91]: TypeError is the more appropriate type

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError

File: src/crawler.js [hunk 5]
```diff
@@ -71,22 +75,43 @@ function normalizeIpp(ip, port) {
...
 
 Crawler.prototype.getCrawl = function(entryIp) {
   var self = this;
-  return new Promise(function(resolve, reject){
+  return new Promise(function(resolve, reject) {
+    if (entryIp === undefined) {
+      return reject(new Error("Invalid ip address")) 
+    }
+    if (entryIp.split(' ').length != 1 || !IPP_PATTERN.test(entryIp)) {
+      return reject(new Error("Invalid ip address (perhaps port missing)"))
```
Comment (tdfischer) [inline src/crawler.js]: You can just `throw new Error` without having to worry about calling reject(). It also speeds up execution by a little bit since it doesn't have an extra call to reject().

https://stackoverflow.com/questions/28703241/promise-reject-vs-throw-error

Having an explicit throw in there would also mean you wouldn't need the return statement, and it'd make the code absolutely explicit that it has encountered a failure situation.

File: src/crawler.js [hunk 6]
```diff
@@ -112,7 +137,7 @@ Crawler.prototype.crawl = function(ipp, hops) {
     self.dequeue(ipp);
 
     if (err) {
-      self.logger.error(ipp + ' has err ', err);
+      //self.logger.error(ipp + ' has err ', err);
```
Comment (tdfischer) [inline src/crawler.js]: Its best to strip out commented code or document why it is commented out, since future hackers will read this and wonder why it was commented out in the first place.

File: test/crawl-test.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
+var promise = require('chai').expect
+var assert = require('chai').assert
+var expect = require('chai').expect
+var _ = require('lodash');
+var Crawler = require('../src/crawler.js').Crawler;
+
+describe('Crawler', function() {
+  describe('#constructor', function() {
+    // max requests
+    it('Shouldn\'t return an error when given valid max requests', function() {
```
Comment (tdfischer) [inline test/crawl-test.js]: Double quotes can eliminate the need to escape single quotes in these strings.

I think the phrasing you're looking for in these tests is "Shouldn't throw an error" instead of "return". Hackers are real pedantic about stuff like that :(

File: test/crawl-test.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      var maxRequests = "lwarc";
+      expect(function() {new Crawler(maxRequests)}).to.throw(Error);
+    });
+
+    // logger
+    it('Shouldn\'t return an error when given valid logger', function() {
+      var crawler = new Crawler(100, console);
+    });
+    it('Shouldn\'t return an error when given undefined logger', function() {
+      var crawler = new Crawler(100, undefined);
```
Comment (tdfischer) [inline test/crawl-test.js]: Just to be extra sure, this should be `new Crawler(100)`, as in drop the second argument entirely.

File: test/crawl-test.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      return crawler.getCrawl(entryIP).catch(function(error) {
+          assert.strictEqual(error.message, "Invalid ip address (perhaps port missing)")
+      })
+      .then(done, done);
+    });
+    it('Should return an error when given ip address without port', function(done) {
+      var crawler = new Crawler(100)
+      var entryIP = "212.83.147.166"
+      return crawler.getCrawl(entryIP).catch(function(error) {
+          assert.strictEqual(error.message, "Invalid ip address (perhaps port missing)")
```
Comment (tdfischer) [inline test/crawl-test.js:62]: I wonder if the API should be changed to `getCrawl(entryIP, entryPort)`. Then you wouldn't need to worry about parsing any strings for ports.

Probably fine for right now until API usage bumps up against it. Something to keep in mind for the future.

File: src/crawler.js (hunk 1 above)
Comment (sourencho) [inline src/crawler.js:73]: if ip is undefined, isn't it normal for this function to return undefined also?
it's up to the caller to not pass in an undefined ip

File: test/crawl-test.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      this.timeout(10000)
+      var crawler = new Crawler(100)
+      var entryIP = '192.170.145.70:51235'
+
+      crawler.getCrawl(entryIP).then(function(data) {
+        expect(data).to.have.property('start');
+        expect(data).to.have.property('end');
+        expect(data).to.have.property('entry');
+        expect(data).to.have.property('data');
+        expect(data).to.have.property('errors');
```
Comment (tdfischer) [inline test/crawl-test.js:84]: Can this also test the types of each of these properties? Check if they're arrays, strings, etc.

File: src/crawler.js (hunk 1 above)
Comment (tdfischer) [inline src/crawler.js:73]: I think it'd be better to raise an exception than return undefined.

Raising an exception is a very disruptive thing that forces the API consumer to sit in the corner and think about what they've done since the whole program will explode.

Returning undefined means that the API user needs to remember to check for a valid return value from a function. In this particular instance, a function called `normalizeIpp` implies that it'll return a normalized IP and port in the form of a string. Returning undefined breaks that assumption.

tl;dr Exceptions are explicit error reporting, returning undefined values is implicit error reporting, and programmers really hate having to remember things.

**Comment by tdfischer** [comment]
Just a handful of minor issues, nothing serious. Looks good otherwise.

After issues are resolved and this is merged, we can work on getting test coverage hooked up with coveralls.

File: src/crawler.js (hunk 1 above)
Comment (sublimator) [inline src/crawler.js:73]: Agree, also `The stronger wind, the stronger trees`
